TransformFusion/Logic/vtkSlicerTransformFusionLogic.cxx

```diff
@@ -38,133 +38,381 @@
+    this->FuseQuaternionAverage( paramNode );
+  }
+  else if ( paramNode->GetFusionMode() == vtkMRMLTransformFusionNode::FUSION_MODE_CONSTRAIN_SHAFT_ROTATION )
+  {
+    this->FuseConstrainShaftRotation( paramNode );
+  }
+}
 
 //-----------------------------------------------------------------------------
-void vtkSlicerTransformFusionLogic::QuaternionAverageFusion()
+void vtkSlicerTransformFusionLogic::FuseQuaternionAverage( vtkMRMLTransformFusionNode* paramNode )
 {
-  vtkMRMLLinearTransformNode* outputNode = this->TransformFusionNode->GetOutputTransformNode();
-  if (outputNode == NULL)
+  bool verboseWarnings = true;
+  bool conditionsMetForFusion = ConditionsForFusion( paramNode, verboseWarnings );
+  if ( conditionsMetForFusion == false )
+  {
+    return;
+  }
+
+  vtkMRMLLinearTransformNode* outputNode = paramNode->GetOutputTransformNode();
+  if ( outputNode == NULL )
   {
     return;
   }
 
   // Average quaternion
-  vtkSmartPointer<vtkMatrix4x4> outputMatrix = vtkSmartPointer<vtkMatrix4x4>::New();
-  int numberOfInputs = this->TransformFusionNode->GetNumberOfInputTransformNodes();
-  vtkSmartPointer<vtkMatrix4x4> matrix4x4Pointer = vtkSmartPointer<vtkMatrix4x4>::New();
+  vtkSmartPointer< vtkMatrix4x4 > outputMatrix = vtkSmartPointer< vtkMatrix4x4 >::New();
+  int numberOfInputs = paramNode->GetNumberOfInputTransformNodes();
+  vtkSmartPointer< vtkMatrix4x4 > matrix4x4Pointer = vtkSmartPointer< vtkMatrix4x4 >::New();
 
-  float rotationMatrix[3][3] = {{0}};
-  float averageRotationMatrix[3][3] = {{0}};
-  float singleQuaternion[4] = {0};
-  float averageQuaternion[4] = {0};
+  float rotationMatrix[ 3 ][ 3 ] = { { 0 } };
+  float averageRotationMatrix[ 3 ][ 3 ] = { { 0 } };
+  float singleQuaternion[ 4 ] = { 0 };
+  float averageQuaternion[ 4 ] = { 0 };
 
-  for (int i = 0; i < numberOfInputs; i++)
+  for ( int i = 0; i < numberOfInputs; i++ )
   {
 
-    this->TransformFusionNode->GetInputTransformNode(i)->GetMatrixTransformToParent(matrix4x4Pointer);
+    paramNode->GetNthInputTransformNode( i )->GetMatrixTransformToParent( matrix4x4Pointer );
 
-    for (int row = 0; row < 3; row++)
+    for ( int row = 0; row < 3; row++ )
     {
-      for (int column = 0; column < 3; column++)
+      for ( int column = 0; column < 3; column++ )
       { 
-        rotationMatrix[row][column] = matrix4x4Pointer->GetElement(row,column);
+        rotationMatrix[ row ][ column ] = matrix4x4Pointer->GetElement( row, column );
       }
     }
 
-    vtkMath::Matrix3x3ToQuaternion(rotationMatrix, singleQuaternion);
+    vtkMath::Matrix3x3ToQuaternion( rotationMatrix, singleQuaternion );
 
-    averageQuaternion[0] = averageQuaternion[0] + singleQuaternion[0];
-    averageQuaternion[1] = averageQuaternion[1] + singleQuaternion[1];
-    averageQuaternion[2] = averageQuaternion[2] + singleQuaternion[2];
-    averageQuaternion[3] = averageQuaternion[3] + singleQuaternion[3];
+    averageQuaternion[ 0 ] = averageQuaternion[ 0 ] + singleQuaternion[ 0 ];
```

Comment:
This was the pre-existing implementation in the module, I personally have no applications for quaternion average. I could comment + document that the existing implementation is incorrect, but that it could be updated to a proper solution: http://www.acsu.buffalo.edu/~johnc/ave_quat07.pdf